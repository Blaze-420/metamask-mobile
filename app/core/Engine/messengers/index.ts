import { getAccountsControllerMessenger } from './accounts-controller-messenger';
import type { ControllerMessengerByControllerName } from '../types';
import { getEarnControllerMessenger } from './earn-controller-messenger';

/**
 * The messengers for the controllers that have been.
 */
export const CONTROLLER_MESSENGERS: ControllerMessengerByControllerName = {
  AccountsController: {
    getMessenger: getAccountsControllerMessenger,
  },
  EarnController: {
    getMessenger: getEarnControllerMessenger,
  },
} as const;
