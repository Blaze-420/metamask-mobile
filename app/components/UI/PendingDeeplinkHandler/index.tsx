import React, { useEffect } from 'react';
import { View } from 'react-native';
import { useNavigation } from '@react-navigation/native';
import Engine from '../../../core/Engine';
import DeeplinkManager from '../../../core/DeeplinkManager/SharedDeeplinkManager';
import AppConstants from '../../../core/AppConstants';
import Routes from '../../../constants/navigation/Routes';
import { findRouteNameFromNavigatorState } from '../../../util/general';

/**
 * Screen component that handles pending deeplinks
 * This is a non-visual screen that processes any pending deeplinks
 * when the app is unlocked and not on the lock screen
 */
const PendingDeeplinkHandler = () => {
  const navigation = useNavigation();

  useEffect(() => {
    const processPendingDeeplink = () => {
      const pendingDeeplink = DeeplinkManager.getPendingDeeplink();
      const { KeyringController } = Engine.context;
      const route = findRouteNameFromNavigatorState(
        navigation.dangerouslyGetState().routes,
      );

      if (
        pendingDeeplink &&
        KeyringController.isUnlocked() &&
        route !== Routes.LOCK_SCREEN
      ) {
        DeeplinkManager.expireDeeplink();

        DeeplinkManager.parse(pendingDeeplink, {
          origin: AppConstants.DEEPLINKS.ORIGIN_DEEPLINK,
        });
      }
    };

    processPendingDeeplink();
  }, [navigation]);

  return <View />;
};

export default PendingDeeplinkHandler;
