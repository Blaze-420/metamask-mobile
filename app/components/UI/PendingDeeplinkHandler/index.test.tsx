import React from 'react';
import { render } from '@testing-library/react-native';
import PendingDeeplinkHandler from './';
import DeeplinkManager from '../../../core/DeeplinkManager/SharedDeeplinkManager';
import Engine from '../../../core/Engine';
import Routes from '../../../constants/navigation/Routes';

// Mock the dependencies
jest.mock('@react-navigation/native', () => ({
  useNavigation: () => ({
    dangerouslyGetState: () => ({
      routes: [{ name: 'Home' }],
    }),
  }),
}));

jest.mock('../../../core/DeeplinkManager/SharedDeeplinkManager', () => ({
  getPendingDeeplink: jest.fn(),
  expireDeeplink: jest.fn(),
  parse: jest.fn(),
}));

jest.mock('../../../core/Engine', () => ({
  context: {
    KeyringController: {
      isUnlocked: jest.fn(),
    },
  },
}));

describe('PendingDeeplinkHandler', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render an empty View', () => {
    const { toJSON } = render(<PendingDeeplinkHandler />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('process pending deeplink when app is unlocked and not on lock screen', () => {
    const mockPendingDeeplink = 'metamask://test';
    (DeeplinkManager.getPendingDeeplink as jest.Mock).mockReturnValue(
      mockPendingDeeplink,
    );
    (Engine.context.KeyringController.isUnlocked as jest.Mock).mockReturnValue(
      true,
    );

    render(<PendingDeeplinkHandler />);

    expect(DeeplinkManager.getPendingDeeplink).toHaveBeenCalled();
    expect(DeeplinkManager.expireDeeplink).toHaveBeenCalled();
    expect(DeeplinkManager.parse).toHaveBeenCalledWith(mockPendingDeeplink, {
      origin: 'deeplink',
    });
  });

  it('not process deeplink when app is locked', () => {
    const mockPendingDeeplink = 'metamask://test';
    (DeeplinkManager.getPendingDeeplink as jest.Mock).mockReturnValue(
      mockPendingDeeplink,
    );
    (Engine.context.KeyringController.isUnlocked as jest.Mock).mockReturnValue(
      false,
    );

    render(<PendingDeeplinkHandler />);

    expect(DeeplinkManager.getPendingDeeplink).toHaveBeenCalled();
    expect(DeeplinkManager.expireDeeplink).not.toHaveBeenCalled();
    expect(DeeplinkManager.parse).not.toHaveBeenCalled();
  });

  it('not process deeplink when on lock screen', () => {
    const mockPendingDeeplink = 'metamask://test';
    (DeeplinkManager.getPendingDeeplink as jest.Mock).mockReturnValue(
      mockPendingDeeplink,
    );
    (Engine.context.KeyringController.isUnlocked as jest.Mock).mockReturnValue(
      true,
    );

    // Mock navigation state to be on lock screen
    jest
      // eslint-disable-next-line @typescript-eslint/no-require-imports, @typescript-eslint/no-var-requires
      .spyOn(require('@react-navigation/native'), 'useNavigation')
      .mockReturnValue({
        dangerouslyGetState: () => ({
          routes: [{ name: Routes.LOCK_SCREEN }],
        }),
      });

    render(<PendingDeeplinkHandler />);

    expect(DeeplinkManager.getPendingDeeplink).toHaveBeenCalled();
    expect(DeeplinkManager.expireDeeplink).not.toHaveBeenCalled();
    expect(DeeplinkManager.parse).not.toHaveBeenCalled();
  });

  it('not process deeplink when there is no pending deeplink', () => {
    (DeeplinkManager.getPendingDeeplink as jest.Mock).mockReturnValue(null);
    (Engine.context.KeyringController.isUnlocked as jest.Mock).mockReturnValue(
      true,
    );

    render(<PendingDeeplinkHandler />);

    expect(DeeplinkManager.getPendingDeeplink).toHaveBeenCalled();
    expect(DeeplinkManager.expireDeeplink).not.toHaveBeenCalled();
    expect(DeeplinkManager.parse).not.toHaveBeenCalled();
  });
});
