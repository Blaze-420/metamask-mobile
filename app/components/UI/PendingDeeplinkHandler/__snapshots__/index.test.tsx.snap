// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PendingDeeplinkHandler should render an empty View 1`] = `<View />`;
